[ie/afl] added support for Australian Football League (AFL) websites

subrat-lima opened the pull request:
### This PR contains extractors for the following:
- [AFL Video](https://www.afl.com.au/aflw/video)
- [AFL Podcast](https://www.afl.com.au/aflw/podcasts)
- [AFC Video](https://www.afc.com.au/video)
- [Brisbane Lions Video](https://www.lions.com.au/video)
- [Carlton FC Video](https://www.carltonfc.com.au/video)
- [Carlton FC Podcast](https://www.carltonfc.com.au/podcasts)
- OmnyFM Show: [example](https://omny.fm/shows/the-money-show)

Note: 
- OmnyFM is added in this PR as AFL/Carlton FC podcast are OmnyFM show embeds. 
- [AFC radioshow](https://www.afc.com.au/crowsradioshow) is simply a [Listenr](https://www.listnr.com/) embed. It can be added separately in a different PR.

Partially Fixes #10909 
AFC radioshow support has not been added


<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--

# PLEASE FOLLOW THE GUIDE BELOW

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

Commit:
[ie/afl] added AFLVideoIE extractor

Commit:
[ie/afl] added OmnyFMShow extractor

Commit:
[ie/afl][ie/omnyfm] added AFLPodcastIE and updated OmnyFMShowIE

1. AFLPodcastIE: Added extractor for AFL podcasts
2. OmnyFMShowIE: Updated code to adjust url before download page
   to support various url patterns

Commit:
Merge branch 'master' into ie-afl

Commit:
[ie/afl] added AFCVideoIE

Commit:
[ie/afl] add support for lions.com.au and carltonfc.com.au

made the following changes:
1. added support for lions.com.au videos
1. added support for carltonfc.com.au videos and podcasts

pzhlkj6612 commented on yt_dlp/extractor/afl.py:
```diff
@@ -0,0 +1,161 @@
+    def _real_extract(self, url):
+        display_id = self._match_id(url)
+        webpage = self._download_webpage(url, display_id)
+        video_id = self._search_regex(r'"mediaId"\s*:\s*"(\d+)"', webpage, 'video-id')
```
There are multiple `mediaId` properties on the webpage. You need to find the one that is playing.

subrat-lima replied:
This is a blunder on my end by not cross verifying the data properly. Thanks for checking. I'll make sure this doesnt happen again. I've fixed it.

pzhlkj6612 commented on yt_dlp/extractor/afl.py:
```diff
@@ -0,0 +1,161 @@
+        display_id = self._match_id(url)
+        webpage = self._download_webpage(url, display_id)
+        video_id = self._search_regex(r'"mediaId"\s*:\s*"(\d+)"', webpage, 'video-id')
+        player_id = self._search_regex(r'data-player-id\s*=\s*"(\w+)"', webpage, 'player-id') + '_default'
+        account_id = self._search_regex(r'data-account-id\s*=\s*"(\d+)"', webpage, 'account-id')
```
These two attributes are from the "#VideoModal" div, try:

```python
video_attrs = extract_attributes(get_element_html_by_id('VideoModal', webpage))
# ...

```

subrat-lima replied:
Updated extractor to use `extract_attributes` and `get_element_html_by_id`.

pzhlkj6612 commented on yt_dlp/extractor/omnyfm.py:
```diff
@@ -0,0 +1,70 @@
+
+    def _real_extract(self, url):
+        display_id = self._match_id(url)
+        page_url = 'https://omny.fm/shows/' + display_id
```
Why not use the URL directly?

subrat-lima replied:
There are multiple layouts available for this site where the data is present.
So, I thought its better to handle it in such a way that multiple variations of the urls are supported.
example: all these variations are the same podcast but with different endpoints and page layout.
[variation-1](https://omny.fm/shows/tagged/playlists/podcast/embed?style=cover&image=1&share=1&download=1&description=1&follow=1&playlistimages=1&playlistshare=1&foreground=6d4d8f&background=f5f5f5&highlight=feacd2)
[variation-2](https://omny.fm/shows/tagged)
[variation-3](https://omny.fm/shows/tagged/playlists)
[variation-4](https://omny.fm/shows/tagged/playlists/podcast.rss)

Please let me know if this can be handled in a better way.

pzhlkj6612 replied:
Hmm, you just need the substring after the "shows/"? Let's update `_VALID_URL` :

```diff
-_VALID_URL = r'https?://omny\.fm/shows/(?P<id>[\w-]+)'
+_VALID_URL = r'https?://omny\.fm/shows/(?P<id>[^/]+)'

```

<br>

PS: since we are talking about URLs, directly putting those URLs in the comment is OK and would make things a bit more straightforward.

subrat-lima replied:
Understood. Will keep this in mind. I've updated `_VALID_URL`. Thanks

pzhlkj6612 commented on yt_dlp/extractor/omnyfm.py:
```diff
@@ -0,0 +1,70 @@
+        page_url = 'https://omny.fm/shows/' + display_id
+        webpage = self._download_webpage(page_url, display_id)
+
+        data = json.loads(get_element_by_id('__NEXT_DATA__', webpage))
```
```suggestion
        data = self._search_nextjs_data(webpage, display_id)
```

bashonly replied:
also just a note that in general you should avoid using the `get_element` functions for extracting JSON since they unescape html special characters which can break JSON parsing

subrat-lima replied:
updated the extractor to use `_search_nextjs_data` function for json data.

pzhlkj6612 replied:
> avoid using the `get_element` functions for extracting JSON since they unescape html special characters which can break JSON parsing

Example:

```Python
json.loads(get_element_by_id('__NEXT_DATA__', '''
    <script id="__NEXT_DATA__" type="application/json">
    {
        "foo": "bar&quot;baz"
    }
    </script>
'''))

```

The above code throws `json.decoder.JSONDecodeError: Expecting ',' delimiter: line 3 column 25 (char 35)`.

pzhlkj6612 commented on yt_dlp/extractor/omnyfm.py:
```diff
@@ -0,0 +1,70 @@
+        playlist_count = traverse_obj(data, ('props', 'pageProps', 'program', 'DefaultPlaylist', 'NumberOfClips', {int_or_none}))
+        title = traverse_obj(data, ('props', 'pageProps', 'program', 'Name', {str_or_none}))
+        first_page_data = traverse_obj(data, ('props', 'pageProps', 'clips', {dict}))
+        total_pages = math.ceil(playlist_count / self._PAGE_SIZE)
+
+        return self.playlist_result(InAdvancePagedList(
+            functools.partial(self._entries, org_id, playlist_id, first_page_data),
+            total_pages, self._PAGE_SIZE), playlist_id, title)
```
You can skip the calculation about `total_pages` by using `OnDemandPagedList` .

subrat-lima replied:
updated the extractor to use onDemandPagedList.

bashonly replied:
if we know the total count, we should use `InAdvancePagedList`, though

pzhlkj6612 replied:
> if we know the total count, we should use `InAdvancePagedList`, though

Ah, my bad. I think I've misunderstood the use case of `OnDemandPagedList` .

subrat-lima replied:
Added back `InAdvancePagedList`.

Commit:
[ie/omnyfm] updated extractor to use OnDemandPagedList instead of InAdvancePagedList

Commit:
[ie/omnyfm] updated extractor to use _search_nextjs_data for json data parsing

Commit:
[ie/afl] updated AFCVideoIE to extract video attrs using html extract_attributes function

Commit:
[ie/afl] fixed AFLCFVideoIE incorrect video parsing & split CarltonFCVideoIE

Changes made:
1. AFCVideoIE, CarltonFCVideoIE: fixed the issue to find the correct
   video media id. (thanks to pzhlkj6612a)
2. AFCVideoIE: added support for pages where mediaId is unavailable
   in webpage
3. CarltonFCVideoIE: split from AFCVideoIE because of point 2 as the
   variation is specific to AFCVideo
4. Updated test cases

pzhlkj6612 commented on yt_dlp/extractor/afl.py:
```diff
@@ -0,0 +1,204 @@
+
+class AFLVideoIE(InfoExtractor):
+    IE_NAME = 'afl:video'
+    _VALID_URL = r'https?://(?:www\.)?(?:afl|lions)\.com.au/(?:aflw/)?video/(?P<id>\d+)'
```
```suggestion
    _VALID_URL = r'https?://(?:www\.)?(?:afl|lions)\.com\.au/(?:aflw/)?video/(?P<id>\d+)'
```

subrat-lima replied:
Wondering how the earlier regex even worked in the first place. I've added the fix.

pzhlkj6612 commented on yt_dlp/extractor/afl.py:
```diff
@@ -0,0 +1,204 @@
+
+class AFLPodcastIE(InfoExtractor):
+    IE_NAME = 'afl:podcast'
+    _VALID_URL = r'https?://(?:www\.)?(?:afl|carltonfc)\.com.au/(?:aflw/)?podcasts/(?P<id>[\w-]+)'
```
```suggestion
    _VALID_URL = r'https?://(?:www\.)?(?:afl|carltonfc)\.com\.au/(?:aflw/)?podcasts/(?P<id>[\w-]+)'
```

subrat-lima replied:
Fix added

pzhlkj6612 commented on yt_dlp/extractor/afl.py:
```diff
@@ -0,0 +1,204 @@
+
+class AFCVideoIE(InfoExtractor):
+    IE_NAME = 'afc:video'
+    _VALID_URL = r'https?://(?:www\.)?afc\.com.au/video/(?P<id>\d+)'
```
```suggestion
    _VALID_URL = r'https?://(?:www\.)?afc\.com\.au/video/(?P<id>\d+)'
```

subrat-lima replied:
Fix added

pzhlkj6612 commented on yt_dlp/extractor/afl.py:
```diff
@@ -0,0 +1,204 @@
+        player_id = video_attrs['data-player-id'] + '_default'
+        account_id = video_attrs['data-account-id']
+
+        video_element_html = get_element_html_by_attribute('data-id', display_id, webpage)
+        if video_element_html is None:
+            data = self._download_json(f'https://aflapi.afc.com.au/content/aflc-adel/video/en/{display_id}', display_id)
+            video_id = traverse_obj(data, ('mediaId', {str_or_none}))
+        else:
+            video_id = self._search_regex(r'"mediaId"\s*:\s*"(\d+)"', video_element_html, 'video-id', fatal=False)
```
Is `mediaId` an attribute of the element you extracted beforehand?

```suggestion
        video_element_html = get_element_html_by_attribute('data-id', display_id, webpage)
        if not video_element_html:
            video_data = self._download_json(f'https://aflapi.afc.com.au/content/aflc-adel/video/en/{display_id}', display_id)
        else:
            video_data = extract_attributes(video_element_html)
        video_id = video_data['mediaId']
```

And I don't think we should be "non-fatal" here.

subrat-lima replied:
Both `article` and `a` tags have the `data-id` attribute. So, directly using `extract_attributes` on the `video_element_html` will not work as the `article` element is fetched. However, I've tried to follow what you have suggested and replaced `_search_regex` with `_search_json`.
![Screenshot from 2024-09-20 02-05-25](https://github.com/user-attachments/assets/eebba988-00b4-464b-83f1-655b84f0ac7b)

pzhlkj6612 commented on yt_dlp/extractor/afl.py:
```diff
@@ -0,0 +1,204 @@
+
+class CarltonFCVideoIE(InfoExtractor):
+    IE_NAME = 'carltonfc:video'
+    _VALID_URL = r'https?://(?:www\.)?carltonfc\.com.au/video/(?P<id>\d+)'
```
```suggestion
    _VALID_URL = r'https?://(?:www\.)?carltonfc\.com\.au/video/(?P<id>\d+)'
```

subrat-lima replied:
Fix added

pzhlkj6612 commented on yt_dlp/extractor/omnyfm.py:
```diff
@@ -0,0 +1,64 @@
+class OmnyFMShowIE(InfoExtractor):
+    IE_NAME = 'omnyfm:show'
+    _VALID_URL = r'https?://omny\.fm/shows/(?P<id>[\w-]+)'
+    _EMBED_REGEX = [r'<iframe[^>]+?src=(?:["\'])(?P<url>https?://omny\.fm/shows/.+?)\1']
```
The `\1` reference does not work.

```suggestion
    _EMBED_REGEX = [r'<iframe[^>]+?src=(["\'])(?P<url>https?://omny\.fm/shows/.+?)\1']
```

subrat-lima replied:
Wow! Did not know that was the cause behind EMBED functionality not working. Cant thank enough. I've fixed it

Commit:
[ie/omnyfm] updated _EMBED_REGEX

Commit:
[ie/afl] updated _VALID_URL regex to properly escape dot character

Commit:
[ie/omnyfm] updated _VALID_URL regex pattern for id

Commit:
[ie/omnyfm] added back the InAdvancePagedList function

Commit:
[ie/afl] updated AFCVideoIE & CarltonFCVideoIE mediaId parser

Commit:
[ie/afl] refactored AFCVideoIE

1. added tag option to 'get_element_html_by_attribute' function.
   this removes the need for using '_search_json'
2. swapped 'if not' statement with 'if' for simplification

feedback: pzhlkj6612

subrat-lima commented on yt_dlp/extractor/afl.py:
```diff
@@ -0,0 +1,206 @@
+            video_data = self._download_json(f'https://aflapi.afc.com.au/content/aflc-adel/video/en/{display_id}', display_id)
+        else:
+            video_data = self._search_json(r'data-ui-args\s*=\s*["\']', video_element_html, 'video-id', display_id)
+        video_id = video_data['mediaId']
```
Wow, Did not know about the tag option. Thanks. I've updated the code.

pzhlkj6612 commented:
Your git config appears to be wrong. Open https://patch-diff.githubusercontent.com/raw/yt-dlp/yt-dlp/pull/11035.patch and you will see `From: subrat-lima <subrat.k.lima@subrat.k.lima@protonmail.com>` . You may need to correct it for future use. See [Setting your commit email address - GitHub Docs](https://docs.github.com/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address).

After correcting your email, please do NOT update those existing commits (i.e., force-push), otherwise we would lose all review suggestions.

bashonly commented:
@subrat-lima Thank you for your work on this! I ended up merging #15942 since I had fully reviewed it. I appreciate your work on this nonetheless! I will try to get to some of your other PRs soon as I work through the backlog.
